### PR TITLE
Test for chunking upload 

### PR DESCRIPTION
**Pull request by IljaN**
Test for #198 
Requires https://github.com/owncloud/core/pull/31596 to be merged

### CODE REVIEW

**Comment by patrickjahns** [comment]
@IljaN - please:
- add respective labels 
- assign yourself as being responsible when creating the PR
- add to milestone 

On the content - did you check that the behat tests will transform the upload to chunking? 

Checking core feature files for webdav - we explicitly chunk: https://github.com/owncloud/core/blob/master/tests/acceptance/features/apiWebdav/webdav-related-new-endpoint.feature#L464-L473


something to consider - instead of uploading a large file, create the file during the test run - so people do not have to check out unecessary large github repositories

**Comment by PVince81** [comment]
also, as discussed in the core PR:
- [ ] add a test that does a DELETE on the uploads folder for cancellation, to confirm there is no permission issue

**Comment by IljaN** [comment]
There seems to be no pre-defined step to delete files outside of /$user/files, should i create one? @individual-it, @phil-davis

Maybe an abstraction step for the chunking process should be provided which does the whole process without needing to do it manually like this:
```
  @mailhog
  Scenario: A guest user can upload chunked files
    Given as user "admin"
    And user "user0" has been created
    And user "admin" has created guest user "guest" with email "guest@example.com"
    And the HTTP status code should be "201"
    And user "user0" has created a folder "/tmp"
    And user "user0" has shared folder "/tmp" with user "guest@example.com"
    And guest user "guest" has registered
    When user "guest@example.com" creates a new chunking upload with id "chunking-42" using the API
    And user "guest@example.com" uploads new chunk file "1" with "AAAAA" to id "chunking-42" using the API
    And user "guest@example.com" uploads new chunk file "2" with "BBBBB" to id "chunking-42" using the API
    And user "guest@example.com" uploads new chunk file "3" with "CCCCC" to id "chunking-42" using the API
    And user "guest@example.com" moves new chunk file with id "chunking-42" to "/tmp/myChunkedFile.txt" using the API
    Then as "guest@example.com" the file "/tmp/myChunkedFile.txt" should exist
    When user "guest@example.com" deletes folder "/uploads/chunking-42" using the API
    Then as "guest@example.com" the file "/uploads/chunking-42" should not exist
```

**Comment by IljaN** [comment]
@PVince81 test added

**Comment by phil-davis** [comment]
@IljaN yes, compressing that list of chunking steps would be nice for "routine" chunking tests - it can assume some stuff and do it all. If you want detailed control (upload weird numbers of chunks of different sizes out-of-order,...)  then you can write it out as individual steps.
I will make a PR to provide that.

File: tests/acceptance/features/apiGuests/guests.feature [hunk 1]
```diff
@@ -40,6 +40,38 @@ Feature: Guests
...
+    And the HTTP status code should be "201"
+    And user "user0" has created a folder "/tmp"
+    And user "user0" has shared folder "/tmp" with user "guest@example.com"
+    And guest user "guest" has registered
+    When user "guest@example.com" creates a new chunking upload with id "chunking-42" using the API
+    And user "guest@example.com" uploads new chunk file "1" with "AAAAA" to id "chunking-42" using the API
+    And user "guest@example.com" uploads new chunk file "2" with "BBBBB" to id "chunking-42" using the API
+    And user "guest@example.com" uploads new chunk file "3" with "CCCCC" to id "chunking-42" using the API
+    And user "guest@example.com" moves new chunk file with id "chunking-42" to "/tmp/myChunkedFile.txt" using the API
+    Then as "guest@example.com" the file "/tmp/myChunkedFile.txt" should exist
```
Comment (phil-davis) [inline tests/acceptance/features/apiGuests/guests.feature]: You can also add:
```
And as "user0" the file "/tmp/myChunkedFile.txt" should exist
```
just to be sure it ended up where it should.

File: tests/acceptance/features/apiGuests/guests.feature [hunk 2]
```diff
@@ -40,6 +40,38 @@ Feature: Guests
...
+    And the HTTP status code should be "201"
+    And user "user0" has created a folder "/tmp"
+    And user "user0" has shared folder "/tmp" with user "guest@example.com"
+    And guest user "guest" has registered
+    When user "guest@example.com" creates a new chunking upload with id "chunking-42" using the API
+    And user "guest@example.com" uploads new chunk file "1" with "AAAAA" to id "chunking-42" using the API
+    And user "guest@example.com" uploads new chunk file "2" with "BBBBB" to id "chunking-42" using the API
+    And user "guest@example.com" uploads new chunk file "3" with "CCCCC" to id "chunking-42" using the API
+    And user "guest@example.com" cancels chunking-upload with id "chunking-42" using the API
+    Then the HTTP status code should be "204"
```
Comment (phil-davis) [inline tests/acceptance/features/apiGuests/guests.feature]: If you are feeling like really making sure, you can say:
```
And as "user0" the file "/tmp/myChunkedFile.txt" should not exist
```
to be sure the file did not arrive. (Make sure the HTTP status did not lie)
(it seems odd to me that the cancel of chunking is status ``204`` - but I guess that indicates the cancel was successful)

**Comment by patrickjahns** [comment]
@IljaN 
restarted - even though it is now green for `stable10` - these errors strike me - cc @PVince81 
```
[Fri Jun  1 23:26:33 2018] Could not get node for path: "uploads/guest@example.com/chunking-42/.file" : File with name //chunking-42 could not be located
```

**Comment by PVince81** [comment]
first warning is expected, needs fixing in core at some point.

second one is strange... I'll have a look

**Comment by PVince81** [comment]
- [ ] requires https://github.com/owncloud/core/pull/31618 for master

**Comment by patrickjahns** [comment]
@PVince81 - missing master as the required tests step implementation is in core

**Comment by PVince81** [comment]
@patrickjahns I'm not seeing the second error from your comment https://github.com/owncloud/guests/pull/212#issuecomment-394034965 when running tests locally. Was it from a specific drone run ?

**Comment by PVince81** [comment]
@patrickjahns ok I see the error on Drone: https://drone.owncloud.com/owncloud/guests/18/43 line 190.

Since Drone runs with the daily tarball I should try with that one locally as well.

**Comment by PVince81** [comment]
@patrickjahns I managed to reproduce this, I wasn't looking properly. Using a search for ".file" helped me find it in the local log.

It's a core issue: https://github.com/owncloud/core/issues/31631

Some of our Sabre plugins is trying to clean up custom properties and tries to get the node after it was already moved, so it doesn't find it. Nothing to worry about in the context of uploads, but I mentioned in the ticket that we should check whether a real issue might exist with regular files and custom properties.

**Comment by PVince81** [comment]
added more verification steps as advised by @phil-davis 

master got the acceptance test step so I expect everything to be green on all branches now

**Comment by phil-davis** [comment]
Looks good, assuming CI passes.

**Comment by PVince81** [comment]
daily-master-qa... need to wait for tomorrow has it will contain the fix. stay tuned

**Comment by PVince81** [comment]
restarted failed builds, expecting success with today's daily-master-qa

**Pull request merged**